concrete-core/src/backends/core/private/crypto/glwe/keyswitch.rs

```diff
@@ -0,0 +1,887 @@
+    /// }
+    /// assert_eq!(kbd.ciphertext_iter().count(), 15);
+    /// ```
+    #[allow(dead_code)]
+    pub fn ciphertext_iter(
+        &self,
+    ) -> impl Iterator<Item = GlweCiphertext<&[<Self as AsRefTensor>::Element]>>
+    where
+        Self: AsRefTensor,
+    {
+        self.as_tensor()
+            .subtensor_iter(self.glwe_size.0 * self.poly_size.0)
+            .map(move |sub| GlweCiphertext::from_container(sub.into_container(), self.poly_size))
+    }
+
+    /// Returns an iterator over mutably borrowed `LweCiphertext`.
+    ///
+    /// # Example
+    ///
+    /// ```rust,ignore
+    /// use concrete_core::backends::core::private::crypto::{*, glwe::LweKeyBitDecomposition};
+    /// use concrete_core::backends::core::private::math::tensor::{AsRefTensor, AsMutTensor};
+    /// let mut kbd = LweKeyBitDecomposition::from_container(vec![0 as u8; 150], LweSize(10));
+    /// for mut ciphertext in kbd.ciphertext_iter_mut(){
+    ///     ciphertext.as_mut_tensor().fill_with_element(9);
+    /// }
+    /// assert!(kbd.as_tensor().iter().all(|a| *a == 9));
+    /// assert_eq!(kbd.ciphertext_iter().count(), 15);
+    /// ```
+    #[allow(dead_code)]
+    pub fn ciphertext_iter_mut(
+        &mut self,
+    ) -> impl Iterator<Item = GlweCiphertext<&mut [<Self as AsMutTensor>::Element]>>
+    where
+        Self: AsMutTensor,
+    {
+        let chunks_size = self.glwe_size.0 * self.poly_size.0;
+        let poly_size = self.poly_size;
+        self.as_mut_tensor()
+            .subtensor_iter_mut(chunks_size)
+            .map(move |sub| GlweCiphertext::from_container(sub.into_container(), poly_size))
+    }
+
+    /// Consumes the current key bit decomposition and returns an lwe list.
+    ///
+    /// Note that this operation is super cheap, as it merely rewraps the current container in an
+    /// lwe list structure.
+    ///
+    /// # Example
+    ///
+    /// ```rust,ignore
+    /// use concrete_core::backends::core::private::crypto::{*, glwe::LweKeyBitDecomposition};
+    /// let kbd = LweKeyBitDecomposition::from_container(vec![0 as u8; 150], LweSize(10));
+    /// let list = kbd.into_lwe_list();
+    /// assert_eq!(list.count(), CiphertextCount(15));
+    /// assert_eq!(list.lwe_size(), LweSize(10));
+    /// ```
+    pub fn into_glwe_list(self) -> GlweList<Cont> {
+        GlweList {
+            tensor: self.tensor,
+            rlwe_size: self.glwe_size,
+            poly_size: self.poly_size,
+        }
+    }
+}
+
+fn torus_small_sign_decompose<Scalar>(res: &mut [Scalar], val: Scalar, base_log: usize)
+where
+    Scalar: UnsignedTorus,
+    Scalar::Signed: SignedInteger,
+{
+    let mut tmp: Scalar;
+    let mut carry = Scalar::ZERO;
+    let mut previous_carry: Scalar;
+    let block_bit_mask: Scalar = (Scalar::ONE << base_log) - Scalar::ONE;
+    let msb_block_mask: Scalar = Scalar::ONE << (base_log - 1);
+
+    // compute the decomposition from LSB to MSB (because of the carry)
+    for i in (0..res.len()).rev() {
```

Comment:
why is there this copy paste here? that will not be updated with other decompositions easily.

Reply:
That's a good question indeed. It was copy-pasted yes, but from `lwe/keyswitch.rs`, where it probably should have been destroyed a long time ago. I suppose this is some remnants of a very old version of `concrete`.